### PR TITLE
fix(datasets): answer delta queries through cached single-column views

### LINKED ISSUE #2895: Performance regression in _query_hf_dataset when using delta_timestamps (~40x slower)
**Title:** Performance regression in _query_hf_dataset when using delta_timestamps (~40x slower)

---

## Summary

LeRobot v3.0's `_query_hf_dataset` implementation has a severe performance regression when querying multiple indices (e.g., for action horizons with `delta_timestamps`). The current implementation is **~40x slower** than using `select()`.

## Root Cause

In commit `a4aa3164` ("fix data access bottleneck"), the implementation changed from using `select()` to direct indexing:

```python
# Old approach (fast) - used select()
torch.stack(self.hf_dataset.select(q_idx)[key])

# New approach (slow) - direct indexing
# In _query_hf_dataset, it tries column-first access which loads ENTIRE column
torch.stack(self.hf_dataset[key][indices])  # This loads entire column first (~3s)
# Then falls back to row-first
torch.stack(self.hf_dataset[indices][key])  # Still slow compared to select()
```

The column-first approach `hf_dataset[key][indices]` loads the entire column into memory before indexing, which is extremely slow. Even after falling back to row-first access, direct indexing `hf_dataset[indices][key]` is still ~40x slower than using `select()`.

## Benchmark Results

Running on a parquet-based dataset with `delta_timestamps` for 50-frame action horizon:

```
Benchmarking dataset: JianZhou0420/libero_openvla_LeRobotv3_0
  Num samples: 20
  Action horizon: 50

  FPS: 10
  Total frames: 273465

============================================================
BENCHMARK RESULTS
============================================================
1. hf_dataset[i] (single frame):
   0.152s total, 7.6ms per sample

2. hf_dataset.select(indices)[key] (OpenPi's fast approach):
   0.128s total, 6.4ms per sample

3. hf_dataset[indices][key] (LeRobot v3.0's slow approach):
   5.253s total, 262.6ms per sample

4. LeRobotDataset[i] with delta_timestamps (what training uses):
   6.980s total, 349.0ms per sample

============================================================
ANALYSIS
============================================================
select() is 40.6x faster than direct indexing

Estimated time per epoch with current LeRobotDataset:
  26.5 hours (1591 minutes)

Estimated time per epoch with select() fix:
  0.56 hours (34 minutes)
```

| Method | Time per sample | Relative |
|--------|----------------|----------|
| `hf_dataset.select(indices)[key]` | 6.4ms | 1x (baseline) |
| `hf_dataset[indices][key]` | 262.6ms | 41x slower |
| `LeRobotDataset[i]` with delta_timestamps | 349.0ms | 55x slower |

**Impact:** Training time increases from ~34 minutes to ~26.5 hours per epoch.

## Use Case

This affects Pi0/OpenPi style action chunking where we need to query 50-frame action horizons using `delta_timestamps`. This is a common pattern for diffusion-based action prediction models.

```python
# Pi0/OpenPi action chunking pattern
delta_timestamps = {"actions": [t / fps for t in range(50)]}  # 50-frame horizon
dataset = LeRobotDataset(repo_id, delta_timestamps=delta_timestamps)

# Each __getitem__ call queries 50 indices - very slow with current implementation
sample = dataset[i]
```

## Suggested Fix

Use `select()` for multi-index queries in `_query_hf_dataset`:

```python
# Fast approach
torch.stack(self.hf_dataset.select(indices)[key])
```

## Benchmark Script

<details>
<summary>Click to expand benchmark_dataset_loading.py</summary>

```python
"""Benchmark LeRobotDataset loading speed.

Compares different data access methods to measure performance impact
of LeRobot v3.0's slow _query_hf_dataset implementation.

Usage:
    python scripts/benchmark_dataset_loading.py
    python scripts/benchmark_dataset_loading.py --num-samples 100
    python scripts/benchmark_dataset_loading.py --repo-id your/dataset
"""

import argparse
import time

import torch


def benchmark_hf_dataset_direct(dataset, num_samples: int) -> float:
    """Benchmark direct hf_dataset access (no delta_timestamps processing)."""
    start = time.time()
    for i in range(num_samples):
        _ = dataset.hf_dataset[i]
    return time.time() - start


def benchmark_hf_dataset_select(dataset, num_samples: int, horizon: int = 50) -> float:
    """Benchmark hf_dataset.select() for action horizon queries."""
    start = time.time()
    for i in range(num_samples):
        indices = list(range(i, min(i + horizon, len(dataset))))
        _ = torch.stack(dataset.hf_dataset.select(indices)['actions'])
    return time.time() - start


def benchmark_hf_dataset_direct_indexing(dataset, num_samples: int, horizon: int = 50) -> float:
    """Benchmark hf_dataset[indices][key] (LeRobot v3.0's slow approach)."""
    start = time.time()
    for i in range(num_samples):
        indices = list(range(i, min(i + horizon, len(dataset))))
        _ = torch.stack(dataset.hf_dataset[indices]['actions'])
    return time.time() - start


def benchmark_lerobot_dataset(dataset_with_delta, num_samples: int) -> float:
    """Benchmark LeRobotDataset with delta_timestamps (full __getitem__)."""
    start = time.time()
    for i in range(num_samples):
        _ = dataset_with_delta[i]
    return time.time() - start


def main():
    parser = argparse.ArgumentParser(description="Benchmark LeRobotDataset loading speed")
    parser.add_argument("--repo-id", default="JianZhou0420/libero_openvla_LeRobotv3_0",
                        help="HuggingFace dataset repo ID")
    parser.add_argument("--num-samples", type=int, default=50,
                        help="Number of samples to benchmark")
    parser.add_argument("--horizon", type=int, default=50,
                        help="Action horizon for delta_timestamps")
    args = parser.parse_args()

    print(f"Benchmarking dataset: {args.repo_id}")
    print(f"  Num samples: {args.num_samples}")
    print(f"  Action horizon: {args.horizon}")
    print()

    # Import here to see any warnings
    from lerobot.datasets.lerobot_dataset import LeRobotDataset, LeRobotDatasetMetadata

    # Load metadata
    meta = LeRobotDatasetMetadata(args.repo_id)
    fps = meta.fps
    print(f"  FPS: {fps}")
    print(f"  Total frames: {meta.total_frames}")
    print()

    # Load dataset WITHOUT delta_timestamps
    print("Loading dataset without delta_timestamps...")
    dataset_simple = LeRobotDataset(args.repo_id)

    # Load dataset WITH delta_timestamps
    print("Loading dataset with delta_timestamps...")
    delta_timestamps = {"actions": [t / fps for t in range(args.horizon)]}
    dataset_with_delta = LeRobotDataset(args.repo_id, delta_timestamps=delta_timestamps)

    print()
    print("=" * 60)
    print("BENCHMARK RESULTS")
    print("=" * 60)

    # Benchmark 1: Direct hf_dataset access
    elapsed = benchmark_hf_dataset_direct(dataset_simple, args.num_samples)
    print(f"1. hf_dataset[i] (single frame):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    # Benchmark 2: select() method (fast)
    elapsed = benchmark_hf_dataset_select(dataset_simple, args.num_samples, args.horizon)
    print(f"2. hf_dataset.select(indices)[key] (OpenPi's fast approach):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    # Benchmark 3: Direct indexing (slow - LeRobot v3.0's approach)
    elapsed = benchmark_hf_dataset_direct_indexing(dataset_simple, args.num_samples, args.horizon)
    print(f"3. hf_dataset[indices][key] (LeRobot v3.0's slow approach):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    # Benchmark 4: Full LeRobotDataset with delta_timestamps
    elapsed = benchmark_lerobot_dataset(dataset_with_delta, args.num_samples)
    print(f"4. LeRobotDataset[i] with delta_timestamps (what training uses):")
    print(f"   {elapsed:.3f}s total, {elapsed/args.num_samples*1000:.1f}ms per sample")
    print()

    print("=" * 60)
    print("ANALYSIS")
    print("=" * 60)

    # Calculate speedup
    t_select = benchmark_hf_dataset_select(dataset_simple, 10, args.horizon) / 10
    t_direct = benchmark_hf_dataset_direct_indexing(dataset_simple, 10, args.horizon) / 10
    speedup = t_direct / t_select if t_select > 0 else 0

    print(f"select() is {speedup:.1f}x faster than direct indexing")
    print()

    # Estimate full training impact
    total_frames = meta.total_frames
    t_per_sample = elapsed / args.num_samples
    estimated_epoch_time = total_frames * t_per_sample
    print(f"Estimated time per epoch with current LeRobotDataset:")
    print(f"  {estimated_epoch_time/3600:.1f} hours ({estimated_epoch_time/60:.0f} minutes)")
    print()

    # Estimate with fix
    t_fast = benchmark_hf_dataset_select(dataset_simple, args.num_samples, args.horizon) / args.num_samples
    estimated_fast_epoch = total_frames * t_fast
    print(f"Estimated time per epoch with select() fix:")
    print(f"  {estimated_fast_epoch/3600:.2f} hours ({estimated_fast_epoch/60:.0f} minutes)")


if __name__ == "__main__":
    main()
```

</details>

## Environment

- LeRobot version: v3.0 (latest main)
- Dataset format: parquet (images stored in parquet, not video)
- Python: 3.10
- Platform: Linux

### PR DESCRIPTION
**Pull request by Rick0525**
## Summary / Motivation

Since `datasets` 4.4, setting a custom transform disables the lazy-`Column` fast path, so both `hf_dataset[key][indices]` and `hf_dataset[indices][key]` fetch and decode entire rows. LeRobot always sets `hf_transform_to_torch`, so every delta query pays full-row cost: on image datasets (embedded images in parquet) that means decoding `chunk_size x n_cameras` images per sample just to read a small float column like `action`. Full root-cause analysis with `datasets` source references is in https://github.com/huggingface/lerobot/issues/2895#issuecomment-5060174507 (the guard is unchanged as of `datasets` 5.0.0, so no upgrade fixes it).

This PR routes delta queries through cached per-key `select_columns` views. `select_columns` is a zero-copy schema projection, so row queries on the view fetch and decode only the requested column. The view keeps the `hf_transform_to_torch` transform, which is column-wise, so outputs are unchanged.

Design notes:

- The views rely on schema projection plus index-based row gather, both stable public `datasets` semantics, rather than on which internal path `ds[key][idx]` takes in a given release. This is what made the previous code flip between fast and slow across `datasets` versions (#2408, #2549, #3210, #3523).
- The try/except fallback in `_query_hf_dataset` is no longer needed: the view path is the row-query semantics, on a one-column schema.
- If `datasets` later ships a column-safe transform declaration (discussed with @lhoestq in #2895), the cached views can be replaced by a one-line declaration at `set_transform` time. The workaround has a clear retirement path.

## Related issues

- Fixes: #2895
- Related: #3523 (benchmarks the `datasets` 4.3 to 4.4 cliff), #3558 (orthogonal: hoists deduplicated row-batch fetches; the two compose), #2408 (introduced the column-first attempt), #2549 (the pre-4.0 flavor of this bug)

## What changed

- `DatasetReader._column_view(key)`: builds and caches a single-column `select_columns` view per queried key; the cache is invalidated when `hf_dataset` is reloaded.
- `_query_hf_dataset` answers delta queries through the view instead of try/except between two equally slow full-row paths.
- `_get_query_timestamps` uses the same view for the `timestamp` column.
- No behavior change: outputs are bit-identical (see tests).

## How was this tested (or how to run locally)

- Tests added in `tests/datasets/test_dataset_reader.py`:
  - `test_query_hf_dataset_matches_row_query`: bit-identical (`torch.equal`) against a plain row query, over every frame of an image-type dataset, with a delta window that crosses episode boundaries on both ends (padding/clamping cases).
  - `test_delta_query_transform_receives_only_requested_column`: contract test asserting the transform only ever receives the requested column during delta queries.
- `pytest -q tests/datasets/test_dataset_reader.py`
- Benchmark (lerobot 0.6.0 + this patch backported, datasets 4.8.5, `HuggingFaceVLA/libero`, 273,465 rows, 2 embedded-image cameras, 50-index `action` chunk query, median of 12 interleaved rounds): 165.0 ms (current) to 0.8 ms (this PR), ~200x. End-to-end `__getitem__` per sample: 200.7 ms to 4.4 ms (46x). In a 2xA800 DDP SmolVLA fine-tuning run, dataloader wait dropped from 38% of wall-clock to ~1%.
- Production validation: outputs bit-identical against the old path over 324 samples including 24 episode-boundary cases.

## Checklist (required before merge)

- [x] Linting/formatting run (`pre-commit run -a`)
- [x] All tests pass locally (`pytest`)
- [x] Documentation updated (n/a: no public API change)
- [ ] CI is green
- [x] Community Review: I have reviewed another contributor's open PR and linked it here: #3558

## Reviewer notes

- The one subtle point is transform semantics: the view carries `hf_transform_to_torch`, which processes each column independently (`for key in items_dict` in `io_utils.py`), so feeding it a single-column batch cannot change outputs. The contract test pins this down.
- `_column_view` invalidates by identity check on `hf_dataset`, covering `try_load` / `load_and_activate` reloads. In-place `set_transform` on an already-queried reader is not a supported pattern today; views built afterwards would keep the old transform.

### CODE REVIEW

**Comment by Yongtae723** [comment]
We hit this exact issue in production training (reported as #4392, which turned out to be a duplicate of #2895) and reviewed this PR against our setup. Sharing the results as a data point for merging:

**Production impact of the bug** (embedded-JPEG dataset, 2 cameras, `delta_timestamps` with a 100-step action chunk): each `__getitem__` decoded ~200 images that were immediately discarded — ~300 ms/sample. ACT training throughput was 1.2 it/s with 32 dataloader workers pegged at 100% CPU. After bypassing the row-query path (equivalent to what this PR does), the same training runs at **13.6 it/s** with ~3 CPU cores total.

**Review of this PR:**
- The mechanism is correct for this case: `select_columns` views make delta queries fetch/decode only the requested column, and since `hf_transform_to_torch` is column-wise, outputs are unchanged. We independently verified output equivalence on our dataset (chunk contents and `*_is_pad` bit-identical to the row-query path, including episode-boundary padding).
- One theoretical gap: the view cache is invalidated only on `hf_dataset` **identity** change, but `set_transform()` mutates the dataset in place. A user calling `set_transform` after the first delta query gets cached views holding the *old* transform (your own test hints at this: "Set the spy before the first query"). LeRobot's internal usage sets the transform once at load, so this doesn't bite in practice, but including the transform's identity in the cache key would close it cheaply.

Either way, +1 on rebasing/merging this — the current behavior makes image-dtype datasets effectively unusable with action chunking.

**Review by lhoestq — APPROVED** [review]
lgtm !

**Comment by Rick0525** [comment]
@Yongtae723 Thanks a lot for the thorough review and the production data point. An independent bit-identical equivalence check on a real setup is exactly the validation this needed.

You're right about the `set_transform` gap, and it's real: the dataset mutates in place, so the identity check never fires and a view built before the call keeps answering delta queries with the old transform. Fixed in 79f016f6 as you suggested. The transform recorded at view-build time is now part of the cache key, and views are rebuilt when it changes, with a regression test covering exactly this scenario.

Also rebased onto current main to resolve the conflict with #4182. the branch is mergeable again.

**Comment by Rick0525** [comment]
Gentle ping @CarolinePascal, this seems to fall in your area (datasets), and I'd appreciate a look when you have a chance. Quick status so it's easy to pick up:

- Fixes #2895: since `datasets` 4.4, a custom transform disables the lazy-`Column` fast path, so every delta query decodes full rows. On image-in-parquet datasets that means `chunk_size × n_cameras` images decoded per sample just to read a small float column like `action`.
- @lhoestq has reviewed and approved. @Yongtae723 independently validated the patch on a production setup (bit-identical outputs vs. main, timing numbers in his comment above).
- The only push since that approval fixes a gap his review caught: `set_transform` mutates the dataset in place, so the old identity check never fired and a cached view could keep serving the old transform. Cached views are now invalidated explicitly when the transform changes.
- The branch is conflict-free. The test workflows are still waiting for maintainer approval to run, so approving the CI trigger would also help.

**Comment by CarolinePascal** [comment]
Hi @Rick0525,

Sorry for the delayed answer - The PR looks good, could resolve the conflicts so I can move on with the merge ? 

Thanks !

**Comment by Rick0525** [comment]
@CarolinePascal Done. Rebased onto current main and resolved the conflicts (the new reader-registry tests and this PR's delta-query tests landed in the same spot in `test_dataset_reader.py`). No content changes to either commit beyond the rebase, and `tests/datasets/test_dataset_reader.py` passes locally (15 passed). Thanks!

**Review by CarolinePascal — APPROVED** [review]
LGTM !
